plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/MethodSignatureEmbedding.kt

```diff
@@ -5,31 +5,35 @@
 
 package org.jetbrains.kotlin.formver.embeddings
 
-import org.jetbrains.kotlin.formver.conversion.ReturnVariableName
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.formver.conversion.ResultTrackingContext
+import org.jetbrains.kotlin.formver.conversion.StmtConversionContext
+import org.jetbrains.kotlin.formver.domains.convertType
 import org.jetbrains.kotlin.formver.viper.MangledName
-import org.jetbrains.kotlin.formver.viper.ast.*
+import org.jetbrains.kotlin.formver.viper.ast.Exp
 
 /**
- * This embedding represents a method signature.
+ * This embedding represents a method signature without parameter names.
  * In case the method has a receiver it becomes the first argument of the function.
  * Example: Foo.bar(x: Int) --> Foo$bar(this: Foo, x: Int)
  */
 interface MethodSignatureEmbedding {
     val name: MangledName
-    val receiver: VariableEmbedding?
-    val params: List<VariableEmbedding>
+    val receiverType: TypeEmbedding?
+    val paramTypes: List<TypeEmbedding>
     val returnType: TypeEmbedding
 
-    val returnVar
-        get() = VariableEmbedding(ReturnVariableName, returnType)
-    val formalArgs: List<VariableEmbedding>
-        get() = listOfNotNull(receiver) + params
+    val formalArgTypes: List<TypeEmbedding>
+        get() = listOfNotNull(receiverType) + paramTypes
+}
 
-    fun toMethodCall(
-        parameters: List<Exp>,
-        targetVar: VariableEmbedding,
-        pos: Position = Position.NoPosition,
-        info: Info = Info.NoInfo,
-        trafos: Trafos = Trafos.NoTrafos,
-    ) = Stmt.MethodCall(name, parameters, listOf(targetVar.toLocalVar()), pos, info, trafos)
-}
\ No newline at end of file
+// This is pretty much a mess, but will get better when we have typed expression embeddings.
+fun MethodSignatureEmbedding.callWithConversions(
```

Comment:
I'm not sure I'm a huge fan of this function. For now it's quite unreadable especially with the `call` parameter and once we have typed expression embeddings this function will not just get better but obsolete since then `toMethodCall` has all necessary information to do the conversion.